devtools/startup/DevToolsStartup.jsm

```diff
@@ -1632,237 +1584,14 @@ async function fetchURL(url) {
-    console.error("Unstable recording: " + data.why);
-    const browser = getBrowser();
-
-    reloadAndClearRecordingState(browser, `https://replay.io/browser/error?message=${data.why}`);
-  });
-  recording.on("finished", function(name, data) {
-    pingTelemetry('record-button', 'finished', {...data, recordingId: data.id});
-
-    browser = getBrowser();
-    const recordingId = data.id;
-
-    // When the submitTestRecordings pref is set we don't load the viewer,
-    // but show a simple page that the recording was submitted, to make things
-    // simpler for QA and provide feedback that the pref was set correctly.
-    if (
-      Services.prefs.getBoolPref("devtools.recordreplay.submitTestRecordings")
-    ) {
-      fetch(`https://test-inbox.replay.io/${recordingId}:${browser.currentURI.spec}`);
-      const why = `Test recording added: ${recordingId}`;
-      reloadAndClearRecordingState(browser, `about:replay?submitted=${why}`);
-      return;
-    }
-
-    reloadAndClearRecordingState(browser);
-
-    recordReplayLog(`FinishedRecording ${recordingId}`);
-  });
-
-  recording.on("saved", function(name, data) {
-    pingTelemetry('record-button', 'saved', {...data, recordingId: data.id});
-
-    const recordingId = data.id;
-
-    // suppress launching new tab for test recordings
-    if (Services.prefs.getBoolPref("devtools.recordreplay.submitTestRecordings")) {
-      return;
-    }
-
-    try {
-      // Find the dispatcher to connect to.
-      const dispatchAddress = getDispatchServer();
-
-      let extra = "";
-
-      // Specify the dispatch address if it is not the default.
-      if (dispatchAddress != "wss://dispatch.replay.io") {
-        extra += `&dispatch=${dispatchAddress}`;
-      }
-
-      // For testing, allow specifying a test script to load in the tab.
-      const localTest = env.get("RECORD_REPLAY_LOCAL_TEST");
-      if (localTest) {
-        extra += `&test=${localTest}`;
-      } else if (!isAuthenticationEnabled()) {
-        // Adding this urlparam disables checks in the devtools that the user has
-        // permission to view the recording.
-        extra += `&test=1`;
-      }
-
-      const tabbrowser = browser.getTabBrowser();
-      const currentTabIndex = tabbrowser.visibleTabs.indexOf(tabbrowser.selectedTab);
-      const tab = tabbrowser.addTab(
-        `${getViewURL()}?id=${recordingId}${extra}`,
-        { triggeringPrincipal, index: currentTabIndex === -1 ? undefined : currentTabIndex + 1}
-      );
-      tabbrowser.selectedTab = tab;
-    } catch (e) {
-      pingTelemetry('record-button', 'save-error', {...data, recordingId: data.id, error: e});
-    }
-
-    recordReplayLog(`SavedRecording ${recordingId}`);
-  });
-}
-
 Services.obs.addObserver(
-  subject => onRecordingStarted(subject.wrappedJSObject),
-  "recordreplay-recording-started"
+  subject => refreshRecordingButton(subject.wrappedJSObject.browser.ownerDocument),
+  "recordreplay-recording-changed"
 );
```

Comment:
If you feel like, it seems like it would be more consistent to change this to be `setRecordingStateChangeCallback` on the connection directly, like we do for `setConnectionStatusChangeCallback`, then we don't need to rely any observers. This may have never needed to be an observer to begin with, but I don't remember where the state of the code was when I added this.

Reply:
That's fine with me. I don't mind the event listener approach as it does provide more built in extensibility than setting a callback. Happy to change this out if you think it's best here.

Reply:
I think it's nicer to have the callback directly rather than relying on a global event bus, but also an example of something we can always change later and does not need to be a blocker.